packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,182 @@
+    this.processor.end(new ErrorAborted("BackfillSync"));
+  }
+
+  async init(): Promise<void> {
+    const anchorState = this.chain.getHeadState();
+    const parentBlockRoot: Root = anchorState.latestBlockHeader.parentRoot;
+    this.anchorBlock = await this.db.blockArchive.getByRoot(parentBlockRoot);
+    void this.sync();
+  }
+
+  async sync(): Promise<void> {
+    try {
+      for await (const _ of this.processor) {
+        if (this.peers.size === 0) continue;
+
+        const oldestRequiredEpoch = Math.max(
+          GENESIS_EPOCH,
+          this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+        );
+        const oldestSlotRequired = computeStartSlotAtEpoch(this.config, oldestRequiredEpoch);
+
+        if (this.anchorBlock) {
+          try {
+            const toSlot = this.lastFetchedSlot ?? this.anchorBlock.message.slot;
+            const fromSlot = Math.max(toSlot - this.opts.batchSize, oldestSlotRequired);
+            if (fromSlot >= toSlot) {
+              this.logger.info("Backfill sync completed");
+              break;
+            }
+            await this.syncRange(fromSlot, toSlot, this.anchorBlock.message.parentRoot);
+          } catch (e) {
+            this.logger.debug("Error while backfiling by range", e);
+          }
+        } else {
+          try {
+            await this.syncBlock(this.chain.getHeadState().latestBlockHeader.parentRoot);
+          } catch (e) {
+            this.logger.debug("Error while backfilling anchor block", e);
+          }
+        }
+      }
+    } catch (e) {
+      if (e instanceof ErrorAborted) {
+        return; // Ignore
+      }
+      this.logger.error("BackfillSync Error", e);
+    }
+  }
+
+  private async addPeer(peerId: PeerId, peerStatus: phase0.Status): Promise<void> {
+    const requiredSlot = this.anchorBlock?.message.slot ?? this.chain.getHeadState().slot;
+    if (peerStatus.headSlot >= requiredSlot) {
+      this.peers.set(peerId, peerStatus);
+      this.processor.trigger();
+    }
+  }
+
+  /**
+   * Remove this peer from all sync chains
+   */
+  private removePeer(peerId: PeerId): void {
+    this.peers.delete(peerId);
+  }
+
+  private async syncBlock(root: Root): Promise<void> {
+    const peer = getRandomPeer(this.peers.keys());
+    try {
+      const blocks = await this.network.reqResp.beaconBlocksByRoot(peer, [root] as List<Root>);
+      if (blocks.length === 0) {
+        throw new BackfillSyncError({code: BackfillSyncErrorCode.MISSING_BLOCKS, peerId: peer});
+      }
+      const block = blocks[0] as SignedBeaconBlock;
+      try {
+        verifyBlockSignature(this.config, this.chain.getHeadState(), block);
+      } catch (e) {
+        throw new BackfillSyncError({code: BackfillSyncErrorCode.INVALID_SIGNATURE});
+      }
+
+      await this.db.blockArchive.put(block.message.slot, block);
+      this.anchorBlock = block;
+      this.processor.trigger();
+    } catch (e) {
+      this.peers.delete(peer);
```

Comment:
You should report the peer and expect the network to disconnect the peer eventually once it is banned. This approach may be too punishing and overlaps with the scoring system.